File: services/client/src/protocol/protocol.go
package protocol

import (
	"encoding/binary"
	"io"
	"strings"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

const (
	MsgBet     byte = 1
	MsgFinish  byte = 2
	MsgWinners byte = 3
)

const (
	msgTypeSize   = 1
	msgLengthSize = 4
	headerSize    = msgTypeSize + msgLengthSize
)

func SendMessage(conn io.Writer, msgType byte, payload []byte) error {
	message := make([]byte, headerSize+len(payload))
	message[0] = msgType
	binary.BigEndian.PutUint32(message[msgTypeSize:headerSize], uint32(len(payload)))
	copy(message[headerSize:], payload)

	return safe_socket.SendAll(conn, message)
}

func ReadMessage(conn io.Reader) (byte, []byte, error) {
	header, err := safe_socket.RecvAll(conn, headerSize)
	if err != nil {
		return 0, nil, err
	}

	msgType := header[0]
	length := binary.BigEndian.Uint32(header[msgTypeSize:])

	payload, err := safe_socket.RecvAll(conn, int(length))
	if err != nil {
		return 0, nil, err
	}

	return msgType, payload, nil
}

func SerializeBet(agencyId, line string) []byte {
	return []byte(agencyId + "," + line)
}

func ParseWinners(payload string) []string {
	if payload == "" {
		return []string{}
	}
	return strings.Split(payload, ",")
}

const documentFieldIndex = 2 // nombre,apellido,documento,nacimiento,numero

func DocumentFromLine(line string) string {
	fields := strings.Split(line, ",")
	return fields[documentFieldIndex]
}

File: services/server/src/protocol/protocol.py
import safe_socket
from lottery import Bet

MSG_BET = 1
MSG_FINISH = 2
MSG_WINNERS = 3

MSG_TYPE_SIZE = 1
MSG_LENGTH_SIZE = 4
HEADER_SIZE = MSG_TYPE_SIZE + MSG_LENGTH_SIZE


def send_message(sock, msg_type, payload):
    header = bytes([msg_type]) + len(payload).to_bytes(MSG_LENGTH_SIZE, "big")
    message = header + payload
    safe_socket.send_all(sock, message)


def recv_message(sock):
    header = safe_socket.recv_all(sock, HEADER_SIZE)
    msg_type = header[0]
    length = int.from_bytes(header[MSG_TYPE_SIZE:], "big")
    payload = safe_socket.recv_all(sock, length)
    return msg_type, payload


def parse_bets(payload):
    text = payload.decode("utf-8")
    bets = []
    for line in text.split("\n"):
        agency_id, first_name, last_name, document, birthdate, number = line.split(",")
        bets.append(Bet(int(agency_id), first_name, last_name, int(document), birthdate, int(number)))
    return bets


def encode_winners(bets):
    documents = [str(bet.document) for bet in bets]
    return ",".join(documents).encode("utf-8")
